app/services/billing-accounts/change-address.service.js

```diff
@@ -0,0 +1,371 @@
+      'agentCompanyId',
+      'contactId',
+      'endDate',
+      'dateUpdated'
+    ])
+}
+
+/**
+ * Persist the address entered during the change address process
+ *
+ * If the address has an `addressId:` we assume it was an existing address selected during the journey. So, the address
+ * is already persisted hence we just return `address`.
+ *
+ * Else we attempt to insert a new address record. If the address has a `uprn:` it will be one selected from the
+ * address lookup page. The previous team also added a unique constraint on UPRN in the table so we cannot insert 2
+ * records with matching UPRNs. But we use this to our advantage. Using `onConflict()` and `merge()` we can have
+ * Objection JS update the existing address record if a matching UPRN exists.
+ *
+ * Because either INSERT or UPDATE gets fired `returning()` will kick in and return the all important `addressId` which
+ * we'll need later for the invoice account address. It will also return the fields specified in the INSERT/UPDATE hence
+ * we get a 'complete' address back that we can return to the calling function.
+ *
+ * > NOTE: We are aware that the existing DB design means multiple billing accounts may refer to the same address
+ * > record. By updating an address record we could be updating the address for multiple records. This is why the
+ * > existing DB design is _bad_! Ideally, billing accounts should have their own address records and we don't worry
+ * > about duplication. This then avoids the problem. But until we can amend the DB design it is better that where a
+ * > record in our DB is locked to OS Places, it should reflect whatever OS Places currently returns, not what it
+ * > returned when first added.
+ */
+async function _persistAddress (trx, address) {
+  if (address.addressId) {
+    return address
+  }
+
+  return address.$query(trx)
+    .insert()
+    .onConflict('uprn')
+    // If a conflict is found this specifies what fields should get updated
+    .merge([
+      'address1',
+      'address2',
+      'address3',
+      'address4',
+      'town',
+      'county',
+      'country',
+      'postcode',
+      'dateUpdated'
+    ])
+    .returning([
+      'addressId'
+    ])
+}
+
+/**
+ * Persist the company (Agent) entered during the change address process
+ *
+ * If the company has a `companyId:` we assume it was an existing company selected during the journey. So, the company
+ * is already persisted hence we just return `company`. If the company name is not set then the user has opted not to
+ * change the agent in the journey. So, we just return the empty `CompanyModel` instance.
+ *
+ * Else we attempt to insert a new company record. If the company has a `companyNumber:` it will either be an existing
+ * company record selected by the user, or they will have been required to enter the company number. The previous team
+ * added a unique constraint on `company_number` in the table so we cannot insert 2 records records with matching
```

Comment:
It might just be my bad English skills but do we need records twice in this sentence?